File: src/encryption.test.ts
import { describe, expect, it } from 'vitest'
import { createEnvelope, generateEd25519KeypairBase64Url, type ProtocolBody } from './protocol.js'
import {
  decryptBody,
  encryptBody,
  generateX25519KeyPair,
  openEnvelope,
  sealEnvelope,
} from './encryption.js'

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function makeEnvelope(body: ProtocolBody) {
  return createEnvelope({
    sender: 'agent-alice',
    recipient: 'agent-bob',
    intent: body.intent,
    content: body.content,
    capability: body.capability,
    correlationId: body.correlationId,
    ttlMs: 60_000,
  })
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

describe('generateX25519KeyPair', () => {
  it('returns 32-byte base64url keys (~43 chars, no padding)', () => {
    const kp = generateX25519KeyPair()
    // base64url of 32 bytes = ceil(32 * 4/3) = 43 chars (no padding)
    expect(kp.publicKey).toMatch(/^[A-Za-z0-9_-]{43}$/)
    expect(kp.privateKey).toMatch(/^[A-Za-z0-9_-]{43}$/)
    // Decoded must be exactly 32 bytes
    expect(Buffer.from(kp.publicKey, 'base64url').length).toBe(32)
    expect(Buffer.from(kp.privateKey, 'base64url').length).toBe(32)
  })
})

describe('sealEnvelope + openEnvelope', () => {
  it('round-trip recovers original body exactly', async () => {
    const recipientKp = generateX25519KeyPair()
    const senderEd = await generateEd25519KeypairBase64Url()

    const originalBody: ProtocolBody = {
      intent: 'TASK',
      content: 'Hello encrypted world!',
      capability: 'some-cap',
      correlationId: 'corr-123',
    }
    const envelope = makeEnvelope(originalBody)

    const sealed = await sealEnvelope(envelope, {
      recipientX25519PublicKey: recipientKp.publicKey,
      senderEd25519PrivateKey: senderEd.privateKey,
    })

    const { body } = await openEnvelope(sealed, {
      recipientX25519PrivateKey: recipientKp.privateKey,
      senderEd25519PublicKey: senderEd.publicKey,
    })

    expect(body.intent).toBe(originalBody.intent)
    expect(body.content).toBe(originalBody.content)
    expect(body.capability).toBe(originalBody.capability)
    expect(body.correlationId).toBe(originalBody.correlationId)
  })

  it('fails with wrong recipient private key (AEAD tag mismatch)', async () => {
    const recipientKp = generateX25519KeyPair()
    const wrongKp = generateX25519KeyPair()
    const senderEd = await generateEd25519KeypairBase64Url()

    // Sign with correct recipient public key but try to open with wrong private key
    // To bypass the signature check we need to seal with wrongKp pubkey too,
    // but the spec says "wrong recipient key" meaning sealed to correct pub key but opened with wrong priv.
    const envelope = makeEnvelope({ intent: 'PING', content: 'secret' })
    const sealed = await sealEnvelope(envelope, {
      recipientX25519PublicKey: recipientKp.publicKey,
      senderEd25519PrivateKey: senderEd.privateKey,
    })

    // We must tamper with the encrypted content to force the wrong key path,
    // but the signature would fail. Instead, bypass openEnvelope and call decryptBody directly
    // with the wrong key — which is what the spec tests.
    expect(() => decryptBody(sealed.body.content, wrongKp.privateKey)).toThrow()
  })

  it('fails if envelope signature is tampered', async () => {
    const recipientKp = generateX25519KeyPair()
    const senderEd = await generateEd25519KeypairBase64Url()

    const envelope = makeEnvelope({ intent: 'PING', content: 'secret' })
    const sealed = await sealEnvelope(envelope, {
      recipientX25519PublicKey: recipientKp.publicKey,
      senderEd25519PrivateKey: senderEd.privateKey,
    })

    // Tamper with the signature value
    const tampered = {
      ...sealed,
      signature: {
        ...sealed.signature!,
        value: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA',
      },
    }

    await expect(
      openEnvelope(tampered, {
        recipientX25519PrivateKey: recipientKp.privateKey,
        senderEd25519PublicKey: senderEd.publicKey,
      }),
    ).rejects.toThrow('signature verification failed')
  })

  it('fails if ciphertext is tampered (AEAD auth tag fails)', async () => {
    const recipientKp = generateX25519KeyPair()
    const senderEd = await generateEd25519KeypairBase64Url()

    const envelope = makeEnvelope({ intent: 'PING', content: 'secret' })
    const sealed = await sealEnvelope(envelope, {
      recipientX25519PublicKey: recipientKp.publicKey,
      senderEd25519PrivateKey: senderEd.privateKey,
    })

    // Decode the encrypted payload, flip a bit in ciphertext, re-encode
    const payloadJson = Buffer.from(sealed.body.content, 'base64url').toString('utf8')
    const payload = JSON.parse(payloadJson) as { ephemeralPublic: string; nonce: string; ciphertext: string; tag: string }
    const ctBuf = Buffer.from(payload.ciphertext, 'base64url')
    ctBuf[0] ^= 0xff  // flip bits in first byte
    const tamperedPayload = { ...payload, ciphertext: ctBuf.toString('base64url') }
    const tamperedContent = Buffer.from(JSON.stringify(tamperedPayload), 'utf8').toString('base64url')

    // We need to re-sign with the same key for the tamper to get past sig verification
    // Instead: call decryptBody directly with tampered content
    expect(() => decryptBody(tamperedContent, recipientKp.privateKey)).toThrow()
  })

  it('two sealEnvelope calls on same body produce different ciphertexts (ephemeral randomness)', async () => {
    const recipientKp = generateX25519KeyPair()
    const senderEd = await generateEd25519KeypairBase64Url()

    const envelope1 = makeEnvelope({ intent: 'PING', content: 'same content' })
    const envelope2 = makeEnvelope({ intent: 'PING', content: 'same content' })

    const sealed1 = await sealEnvelope(envelope1, {
      recipientX25519PublicKey: recipientKp.publicKey,
      senderEd25519PrivateKey: senderEd.privateKey,
    })
    const sealed2 = await sealEnvelope(envelope2, {
      recipientX25519PublicKey: recipientKp.publicKey,
      senderEd25519PrivateKey: senderEd.privateKey,
    })

    // Different ephemeral keys + different nonces → different ciphertexts
    expect(sealed1.body.content).not.toBe(sealed2.body.content)
  })

  it('encrypted content is opaque (does not contain original body.content as plaintext)', async () => {
    const recipientKp = generateX25519KeyPair()
    const senderEd = await generateEd25519KeypairBase64Url()

    const originalContent = 'super-secret-data-12345'
    const envelope = makeEnvelope({ intent: 'TASK', content: originalContent })

    const sealed = await sealEnvelope(envelope, {
      recipientX25519PublicKey: recipientKp.publicKey,
      senderEd25519PrivateKey: senderEd.privateKey,
    })

    // The encrypted content blob should not contain the original plaintext string
    const encryptedContentDecoded = Buffer.from(sealed.body.content, 'base64url').toString('utf8')
    expect(encryptedContentDecoded).not.toContain(originalContent)
    // Also raw base64url should not contain it
    expect(sealed.body.content).not.toContain(originalContent)
  })

  it('encrypted envelope body has correct structure', async () => {
    const recipientKp = generateX25519KeyPair()
    const senderEd = await generateEd25519KeypairBase64Url()

    const envelope = makeEnvelope({ intent: 'PING', content: 'test' })
    const sealed = await sealEnvelope(envelope, {
      recipientX25519PublicKey: recipientKp.publicKey,
      senderEd25519PrivateKey: senderEd.privateKey,
    })

    expect(sealed.body.intent).toBe('ENCRYPTED')
    expect(sealed.body.capability).toBe('x25519-chacha20poly1305')
    expect(sealed.signature).toBeDefined()
    expect(sealed.signature?.alg).toBe('ED25519')
  })
})

describe('encryptBody + decryptBody', () => {
  it('round-trips a body with all optional fields', () => {
    const kp = generateX25519KeyPair()
    const body: ProtocolBody = { intent: 'RESULT', content: 'data', capability: 'cap', correlationId: 'cid-42' }
    const { encryptedContent } = encryptBody(body, kp.publicKey)
    const decrypted = decryptBody(encryptedContent, kp.privateKey)
    expect(decrypted).toEqual(body)
  })
})

File: sdk/go/encryption_test.go
package protocol7h3_test

import (
	"encoding/base64"
	"encoding/json"
	"strings"
	"testing"
	"time"

	protocol7h3 "github.com/IceMasterT/7h3-protocol/sdk/go"
)

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func makeEnvelopeGo(t *testing.T, body protocol7h3.ProtocolBody) protocol7h3.ProtocolEnvelope {
	t.Helper()
	nowMs := time.Now().UnixMilli()
	return protocol7h3.ProtocolEnvelope{
		Header: protocol7h3.ProtocolHeader{
			Version:     "7h3/0.1",
			MessageID:   "msg-go-test",
			TimestampMs: nowMs,
			TTLMs:       60_000,
			Sender:      "agent-alice",
			Recipient:   "agent-bob",
			Nonce:       "test-nonce-go",
		},
		Body: body,
	}
}

// ---------------------------------------------------------------------------
// Test 1: GenerateX25519KeyPair returns 32-byte base64url keys (~43 chars, no padding)
// ---------------------------------------------------------------------------

func TestGenerateX25519KeyPair(t *testing.T) {
	pub, priv, err := protocol7h3.GenerateX25519KeyPair()
	if err != nil {
		t.Fatalf("GenerateX25519KeyPair error: %v", err)
	}

	// Should not contain base64 padding
	if strings.Contains(pub, "=") {
		t.Errorf("public key has padding: %s", pub)
	}
	if strings.Contains(priv, "=") {
		t.Errorf("private key has padding: %s", priv)
	}

	// Should be 43 chars (base64url of 32 bytes without padding)
	if len(pub) != 43 {
		t.Errorf("public key length = %d, want 43", len(pub))
	}
	if len(priv) != 43 {
		t.Errorf("private key length = %d, want 43", len(priv))
	}

	// Decoded must be exactly 32 bytes
	pubBytes, err := base64.RawURLEncoding.DecodeString(pub)
	if err != nil {
		t.Fatalf("decode public key: %v", err)
	}
	if len(pubBytes) != 32 {
		t.Errorf("public key decoded length = %d, want 32", len(pubBytes))
	}

	privBytes, err := base64.RawURLEncoding.DecodeString(priv)
	if err != nil {
		t.Fatalf("decode private key: %v", err)
	}
	if len(privBytes) != 32 {
		t.Errorf("private key decoded length = %d, want 32", len(privBytes))
	}
}

// ---------------------------------------------------------------------------
// Test 2: SealEnvelope + OpenEnvelope round-trip recovers original body exactly
// ---------------------------------------------------------------------------

func TestSealOpenEnvelopeRoundTrip(t *testing.T) {
	recipientPub, recipientPriv, err := protocol7h3.GenerateX25519KeyPair()
	if err != nil {
		t.Fatalf("GenerateX25519KeyPair: %v", err)
	}

	senderPub, senderPriv, err := protocol7h3.GenerateKeypair()
	if err != nil {
		t.Fatalf("GenerateKeypair: %v", err)
	}

	originalBody := protocol7h3.ProtocolBody{
		Intent:        "TASK",
		Content:       "Hello encrypted world!",
		Capability:    "some-cap",
		CorrelationID: "corr-123",
	}
	envelope := makeEnvelopeGo(t, originalBody)

	sealed, err := protocol7h3.SealEnvelope(envelope, recipientPub, senderPriv)
	if err != nil {
		t.Fatalf("SealEnvelope: %v", err)
	}

	_, body, err := protocol7h3.OpenEnvelope(sealed, recipientPriv, senderPub)
	if err != nil {
		t.Fatalf("OpenEnvelope: %v", err)
	}

	if body.Intent != originalBody.Intent {
		t.Errorf("intent = %q, want %q", body.Intent, originalBody.Intent)
	}
	if body.Content != originalBody.Content {
		t.Errorf("content = %q, want %q", body.Content, originalBody.Content)
	}
	if body.Capability != originalBody.Capability {
		t.Errorf("capability = %q, want %q", body.Capability, originalBody.Capability)
	}
	if body.CorrelationID != originalBody.CorrelationID {
		t.Errorf("correlationId = %q, want %q", body.CorrelationID, originalBody.CorrelationID)
	}
}

// ---------------------------------------------------------------------------
// Test 3: OpenEnvelope fails with wrong recipient key
// ---------------------------------------------------------------------------

func TestOpenEnvelopeWrongRecipientKey(t *testing.T) {
	recipientPub, _, err := protocol7h3.GenerateX25519KeyPair()
	if err != nil {
		t.Fatalf("GenerateX25519KeyPair: %v", err)
	}
	_, wrongPriv, err := protocol7h3.GenerateX25519KeyPair()
	if err != nil {
		t.Fatalf("GenerateX25519KeyPair wrong: %v", err)
	}
	senderPub, senderPriv, err := protocol7h3.GenerateKeypair()
	if err != nil {
		t.Fatalf("GenerateKeypair: %v", err)
	}

	envelope := makeEnvelopeGo(t, protocol7h3.ProtocolBody{Intent: "PING", Content: "secret"})
	sealed, err := protocol7h3.SealEnvelope(envelope, recipientPub, senderPriv)
	if err != nil {
		t.Fatalf("SealEnvelope: %v", err)
	}

	// Wrong recipient private key — we need to test decryptBody path directly.
	// To bypass signature check (which uses senderPub correctly), we test OpenEnvelope
	// with wrong key but correct signature — the AEAD tag will fail.
	// However, OpenEnvelope verifies Ed25519 first using senderPub (which matches),
	// so signature passes, and then decryption fails with wrong key.
	_, _, err = protocol7h3.OpenEnvelope(sealed, wrongPriv, senderPub)
	if err == nil {
		t.Error("expected error with wrong recipient key, got nil")
	}
}

// ---------------------------------------------------------------------------
// Test 4: OpenEnvelope fails if envelope signature tampered
// ---------------------------------------------------------------------------

func TestOpenEnvelopeTamperedSignature(t *testing.T) {
	recipientPub, recipientPriv, err := protocol7h3.GenerateX25519KeyPair()
	if err != nil {
		t.Fatalf("GenerateX25519KeyPair: %v", err)
	}
	senderPub, senderPriv, err := protocol7h3.GenerateKeypair()
	if err != nil {
		t.Fatalf("GenerateKeypair: %v", err)
	}

	envelope := makeEnvelopeGo(t, protocol7h3.ProtocolBody{Intent: "PING", Content: "secret"})
	sealed, err := protocol7h3.SealEnvelope(envelope, recipientPub, senderPriv)
	if err != nil {
		t.Fatalf("SealEnvelope: %v", err)
	}

	// Tamper with signature
	tamperedSig := *sealed.Signature
	tamperedSig.Value = strings.Repeat("A", 86)
	sealed.Signature = &tamperedSig

	_, _, err = protocol7h3.OpenEnvelope(sealed, recipientPriv, senderPub)
	if err == nil {
		t.Error("expected error with tampered signature, got nil")
	}
	if !strings.Contains(err.Error(), "signature") {
		t.Errorf("expected signature error, got: %v", err)
	}
}

// ---------------------------------------------------------------------------
// Test 5: OpenEnvelope fails if ciphertext tampered (AEAD auth tag fails)
// ---------------------------------------------------------------------------

func TestOpenEnvelopeTamperedCiphertext(t *testing.T) {
	recipientPub, recipientPriv, err := protocol7h3.GenerateX25519KeyPair()
	if err != nil {
		t.Fatalf("GenerateX25519KeyPair: %v", err)
	}
	senderPub, senderPriv, err := protocol7h3.GenerateKeypair()
	if err != nil {
		t.Fatalf("GenerateKeypair: %v", err)
	}

	envelope := makeEnvelopeGo(t, protocol7h3.ProtocolBody{Intent: "PING", Content: "secret"})
	sealed, err := protocol7h3.SealEnvelope(envelope, recipientPub, senderPriv)
	if err != nil {
		t.Fatalf("SealEnvelope: %v", err)
	}

	// Decode the encrypted payload, flip a bit in ciphertext
	payloadJSON, err := base64.RawURLEncoding.DecodeString(sealed.Body.Content)
	if err != nil {
		t.Fatalf("decode content: %v", err)
	}
	var payload map[string]string
	if err = json.Unmarshal(payloadJSON, &payload); err != nil {
		t.Fatalf("unmarshal payload: %v", err)
	}
	ctBytes, err := base64.RawURLEncoding.DecodeString(payload["ciphertext"])
	if err != nil {
		t.Fatalf("decode ciphertext: %v", err)
	}
	ctBytes[0] ^= 0xFF
	payload["ciphertext"] = base64.RawURLEncoding.EncodeToString(ctBytes)

	tamperedPayloadJSON, _ := json.Marshal(payload)
	tamperedContent := base64.RawURLEncoding.EncodeToString(tamperedPayloadJSON)

	// We need to test decryptBody path, bypassing the signature check by calling
	// a sealed envelope that we signed correctly but with tampered content.
	// Since we can't re-sign (that would require a new seal), test via a helper approach:
	// Create a new sealed envelope with the tampered content re-signed.
	tamperedEnv := sealed
	tamperedEnv.Body.Content = tamperedContent
	// Re-sign it so sig check passes but AEAD fails
	reSealed, err := protocol7h3.SignEnvelopeEd25519(protocol7h3.ProtocolEnvelope{
		Header: tamperedEnv.Header,
		Body:   tamperedEnv.Body,
	}, senderPriv)
	if err != nil {
		t.Fatalf("re-sign tampered envelope: %v", err)
	}

	_, _, err = protocol7h3.OpenEnvelope(reSealed, recipientPriv, senderPub)
	if err == nil {
		t.Error("expected error with tampered ciphertext, got nil")
	}
}

// ---------------------------------------------------------------------------
// Test 6: Two SealEnvelope calls on same body produce different ciphertexts
// ---------------------------------------------------------------------------

func TestSealEnvelopeEphemeralRandomness(t *testing.T) {
	recipientPub, _, err := protocol7h3.GenerateX25519KeyPair()
	if err != nil {
		t.Fatalf("GenerateX25519KeyPair: %v", err)
	}
	_, senderPriv, err := protocol7h3.GenerateKeypair()
	if err != nil {
		t.Fatalf("GenerateKeypair: %v", err)
	}

	body := protocol7h3.ProtocolBody{Intent: "PING", Content: "same content"}
	envelope1 := makeEnvelopeGo(t, body)
	envelope2 := makeEnvelopeGo(t, body)

	sealed1, err := protocol7h3.SealEnvelope(envelope1, recipientPub, senderPriv)
	if err != nil {
		t.Fatalf("SealEnvelope 1: %v", err)
	}
	sealed2, err := protocol7h3.SealEnvelope(envelope2, recipientPub, senderPriv)
	if err != nil {
		t.Fatalf("SealEnvelope 2: %v", err)
	}

	if sealed1.Body.Content == sealed2.Body.Content {
		t.Error("two seal calls produced identical ciphertexts — ephemeral key randomness failure")
	}
}

// ---------------------------------------------------------------------------
// Test 7: Encrypted content is opaque (does not contain original body.content)
// ---------------------------------------------------------------------------

func TestEncryptedContentIsOpaque(t *testing.T) {
	recipientPub, _, err := protocol7h3.GenerateX25519KeyPair()
	if err != nil {
		t.Fatalf("GenerateX25519KeyPair: %v", err)
	}
	_, senderPriv, err := protocol7h3.GenerateKeypair()
	if err != nil {
		t.Fatalf("GenerateKeypair: %v", err)
	}

	originalContent := "super-secret-data-12345"
	envelope := makeEnvelopeGo(t, protocol7h3.ProtocolBody{Intent: "TASK", Content: originalContent})

	sealed, err := protocol7h3.SealEnvelope(envelope, recipientPub, senderPriv)
	if err != nil {
		t.Fatalf("SealEnvelope: %v", err)
	}

	// The raw encrypted content should not contain the original plaintext
	if strings.Contains(sealed.Body.Content, originalContent) {
		t.Error("encrypted content contains original plaintext (raw base64url)")
	}

	// Decoded JSON of encrypted payload should also not contain it
	payloadJSON, _ := base64.RawURLEncoding.DecodeString(sealed.Body.Content)
	if strings.Contains(string(payloadJSON), originalContent) {
		t.Error("encrypted payload JSON contains original plaintext")
	}
}
